### PR TITLE
BirealNet Added

### PR DESCRIPTION
**Pull request by kushagrabhushan**
Implementation of modularised BirealNet code.

### CODE REVIEW

File: trailmet/datasets/classification/cifar.py [hunk 1]
```diff
@@ -190,4 +190,4 @@ def __init__(self, name=None, root=None,
```
Comment (Arnav0400) [inline trailmet/datasets/classification/cifar.py]: Avoid adding these line breaks

File: trailmet/algorithms/binarize/utils.py [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+            best_filename = os.path.join(save, 'tiny-imagenet-model_best.pth.tar')
+            shutil.copyfile(filename, best_filename)
+
+def adjust_learning_rate(optimizer, epoch, args):
+    """Sets the learning rate to the initial LR decayed by 10 every 30 epochs"""
+    lr = args.lr * (0.1 ** (epoch // 30))
+    for param_group in optimizer.param_groups:
+        param_group['lr'] = lr
+        
+def accuracy(output, target, topk1=(1,)):
```
Comment (Arnav0400) [inline trailmet/algorithms/binarize/utils.py]: Can this be a part of trailmet utils?

File: trailmet/algorithms/binarize/utils.py [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+    self.logsoftmax = nn.LogSoftmax(dim=1)
+
+  def forward(self, inputs, targets):
+    log_probs = self.logsoftmax(inputs)
+    targets = torch.zeros_like(log_probs).scatter_(1, targets.unsqueeze(1), 1)
+    targets = (1 - self.epsilon) * targets + self.epsilon / self.num_classes
+    loss = (-targets * log_probs).mean(0).sum()
+    return loss
+
+class AverageMeter(object):
```
Comment (Arnav0400) [inline trailmet/algorithms/binarize/utils.py]: Can both the meter classes be a part of trailmet utils?

File: trailmet/algorithms/binarize/birealnet_model.py [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+
+import torch
+import torch.nn as nn
+import torch.utils.model_zoo as model_zoo
+import torch.nn.functional as F
+import numpy as np
+
+__all__ = ['birealnet18', 'birealnet34']
+
+def conv3x3(in_planes, out_planes, stride=1):
```
Comment (Arnav0400) [inline trailmet/algorithms/binarize/birealnet_model.py]: can you use this function from https://github.com/transmuteAI/trailmet/blob/dev/trailmet/models/resnet.py

File: trailmet/algorithms/binarize/birealnet_model.py [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+        out = self.bn1(out)
+
+        if self.downsample is not None:
+            residual = self.downsample(x)
+
+        out += residual
+
+        return out
+
+class BiRealNet(nn.Module):
```
Comment (Arnav0400) [inline trailmet/algorithms/binarize/birealnet_model.py]: Can this class be inherited from - https://github.com/transmuteAI/trailmet/blob/86baa377e1e52642560db8e44eda0ef2f06a9251/trailmet/models/resnet.py#L151

File: trailmet/algorithms/binarize/birealnet.py [hunk 6]
```diff
@@ -0,0 +1,229 @@
...
+import torch
+import random
+import logging
+import argparse
+import torch.nn as nn
+import torch.utils
+import torch.backends.cudnn as cudnn
+import torch.distributed as dist
+import torch.utils.data.distributed
+sys.path.append('/workspace/code/kushagrabhushan/TrAIL/trailmet/trailmet/algorithms/binarize')
```
Comment (Arnav0400) [inline trailmet/algorithms/binarize/birealnet.py]: Re-work this and make more generic

File: trailmet/algorithms/binarize/birealnet.py [hunk 7]
```diff
@@ -0,0 +1,229 @@
...
+        ## Preparing Optimizers
+        all_parameters = model.parameters()
+        weight_parameters = []
+        for pname, p in model.named_parameters():
+            if p.ndimension() == 4 or pname=='classifier.0.weight' or pname == 'classifier.0.bias':
+                weight_parameters.append(p)
+        weight_parameters_id = list(map(id, weight_parameters))
+        other_parameters = list(filter(lambda p: id(p) not in weight_parameters_id, all_parameters))
+
+        optimizer = torch.optim.Adam(
```
Comment (Arnav0400) [inline trailmet/algorithms/binarize/birealnet.py]: pass the optimizers from kwargs

File: trailmet/algorithms/binarize/utils.py (hunk 2 above)
Comment (kushagrabhushan) [reply trailmet/algorithms/binarize/utils.py]: should i make the save_checkpoint function generalized or is the implementation for the 3 datasets enough?

File: trailmet/algorithms/binarize/utils.py (hunk 3 above)
Comment (kushagrabhushan) [reply trailmet/algorithms/binarize/utils.py]: done

File: trailmet/datasets/classification/cifar.py (hunk 1 above)
Comment (kushagrabhushan) [reply trailmet/datasets/classification/cifar.py]: i dont know what is causing these,  these appeared just by opening and closing the files without changing anything in them.

File: trailmet/algorithms/binarize/birealnet_model.py (hunk 4 above)
Comment (kushagrabhushan) [reply trailmet/algorithms/binarize/birealnet_model.py]: done

File: trailmet/algorithms/binarize/birealnet_model.py (hunk 5 above)
Comment (kushagrabhushan) [reply trailmet/algorithms/binarize/birealnet_model.py]: I directly used the pre-written resnet model and changed the basic block it was using by adding the necessary functions to the resnet.py file. Is that ok?

File: trailmet/algorithms/binarize/birealnet.py (hunk 7 above)
Comment (kushagrabhushan) [reply trailmet/algorithms/binarize/birealnet.py]: done

File: trailmet/algorithms/binarize/birealnet.py (hunk 6 above)
Comment (kushagrabhushan) [reply trailmet/algorithms/binarize/birealnet.py]: Sorry, could you elaborate a little more?

**Pull request closed without merging**